src/pipelinedefinitionstatus.hpp

```diff
@@ -0,0 +1,288 @@
+        return {};
+    }
+    StateChanger<LoadingPreconditionFailedState> handle(const ValidationFailedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateKeeper handle(const UsedModelChangedEvent& e) const {
+        SPDLOG_ERROR("State:{} handling: {}. {}", pipelineDefinitionStateCodeToString(getState()), e.name, INVALID_TRANSITION_MESSAGE);
+        throw std::logic_error(INVALID_TRANSITION_MESSAGE);
+        return {};
+    }
+    StateKeeper handle(const RetireEvent& e) const {
+        SPDLOG_ERROR("State:{} handling: {}. {}", pipelineDefinitionStateCodeToString(getState()), e.name, INVALID_TRANSITION_MESSAGE);
+        throw std::logic_error(INVALID_TRANSITION_MESSAGE);
+        return {};
+    }
+};
+
+struct AvailableState {
+    static const PipelineDefinitionStateCode code = PipelineDefinitionStateCode::AVAILABLE;
+    PipelineDefinitionStateCode getState() const {
+        return code;
+    }
+    void print() const {
+        SPDLOG_ERROR(pipelineDefinitionStateCodeToString(getState()));
+    }
+    StateKeeper handle(const ValidationPassedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<LoadingPreconditionFailedState> handle(const ValidationFailedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<AvailableRequiredRevalidation> handle(const UsedModelChangedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<RetiredState> handle(const RetireEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+};
+
+struct AvailableRequiredRevalidation {
+    static const PipelineDefinitionStateCode code = PipelineDefinitionStateCode::AVAILABLE_REQUIRED_REVALIDATION;
+    PipelineDefinitionStateCode getState() const {
+        return code;
+    }
+    void print() const {
+        SPDLOG_ERROR(pipelineDefinitionStateCodeToString(getState()));
+    }
+    StateChanger<AvailableState> handle(const ValidationPassedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateChanger<LoadingPreconditionFailedState> handle(const ValidationFailedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
+        return {};
+    }
+    StateKeeper handle(const UsedModelChangedEvent& e) const {
+        SPDLOG_INFO("State:{} handling: {}", pipelineDefinitionStateCodeToString(getState()), e.name);
```

Comment:
This line is duplicated in every state change method.  
Maybe we could have one base class for Status and virtual method: `handle`. All other state transition methods could override it and also call base class so we have no code duplication.

Reply:
Maybe an overkill but I wan't to avoid inheritance & cost of it if I can. Those lines most probably can be moved to changeStateTo method tough so I take this point.